File: sem3/map/ToyLanguageInterpreter/src/model/statements/IfStatement.java
package model.statements;

import model.ProgramState;
import model.expressions.Expression;
import utils.ExecutionStack;
import utils.Heap;
import utils.SymbolTable;
import utils.exceptions.InterpreterException;

/**
 * Created by mirko on 18/10/2016.
 */
public class IfStatement implements Statement {
    private Expression expression;
    private Statement ifBranch;
    private Statement elseBranch;

    public IfStatement(Expression expression, Statement ifBranch, Statement elseBranch) {
        this.expression = expression;
        this.ifBranch = ifBranch;
        this.elseBranch = elseBranch;
    }

    @Override
    public ProgramState execute(ProgramState programState) throws InterpreterException {
        ExecutionStack<Statement> executionStack = programState.getExecutionStack();
        SymbolTable<String, Integer> symbolTable = programState.getSymbolTable();
        Heap<Integer, Integer> heap = programState.getHeap();

        if (expression.evaluate(symbolTable, heap) != 0) {
            executionStack.push(ifBranch);
        } else if (elseBranch != null) {
            executionStack.push(elseBranch);
        }

        return programState;
    }

    @Override
    public String toString() {
        StringBuilder string = new StringBuilder();
        string.append("if(" + expression.toString() + ") then(" + ifBranch.toString() + ")");
        if (elseBranch != null) {
            string.append(" else(" + elseBranch.toString() + ")");
        }
        return string.toString();
    }
}

File: sem3/map/ToySharpLanguageInterpreter/ToySharpLanguageInterpreter/src/model/statements/IfStatement.cs
using System;

namespace ToySharpLanguageInterpreter
{
	public class IfStatement : Statement
	{
		private Expression expression;
		private Statement ifBranch;
		private Statement elseBranch;

		public IfStatement(Expression expression, Statement ifBranch, Statement elseBranch)
		{
			this.expression = expression;
			this.ifBranch = ifBranch;
			this.elseBranch = elseBranch;
		}


		public ProgramState execute(ProgramState programState)
		{
			ExecutionStack<Statement> executionStack = programState.getExecutionStack();
			SymbolTable<string, int> symbolTable = programState.getSymbolTable();

			if (expression.evaluate(symbolTable) != 0)
			{
				executionStack.push(ifBranch);
			}
			else if (elseBranch != null)
			{
				executionStack.push(elseBranch);
			}

			return programState;
		}


		public override string ToString()
		{
			string str = "";
			str += "if(" + expression.ToString() + ") then(" + ifBranch.ToString() + ")";
			if (elseBranch != null)
			{
				str += " else(" + elseBranch.ToString() + ")";
			}
			return str;
		}
	}
}